docs/why-fuzz.md

```diff
@@ -0,0 +1,76 @@
+# Why fuzz?
+
+Original authors: Kostya Serebryany, David Drysdale, Chris Lopez, and Max Moroz
+
+This document describes the types of programs that may benefit from [fuzzing](https://en.wikipedia.org/wiki/Fuzzing) and why.
+
+## Introduction
+
+**Fuzz testing is a process of testing APIs with generated data.** The most common forms are:
+- **Mutation based fuzzing** which mutates existing data samples (aka the test corpus) to create test data;
+- **Generation based fuzzing** which produces new test data based on models of the input.
+
+**Guided fuzzing** is an important extension to mutation based fuzzing. Guided fuzzers employ a feedback loop when testing newly mutated inputs. If an input results in a new signal  (such as increased code coverage), it is permanently added to the test corpus. The corpus grows over time, therefore increasing the test coverage of the target program.
+
+**Fuzz testing is valuable for:**
+- Software that receives inputs from untrusted sources **(security)**;
+- Sanity checking the equivalence of two complex algorithms **(correctness)**;
+- Verifying the stability of a high-volume API that takes complex inputs **(stability)**,
+e.g. a decompressor, even if all the inputs are trusted.
+
+Why? **Fuzz testing finds programming errors unrelated to the project requirements.** Errors are always errors, e.g. memory leaks or buffer overflows. Developers will rarely write tests to try and catch these types of bugs, and these are exactly the types of bugs that lead to security holes and reliability problems. Fuzzing is a great and very effective way to find these bugs before they hit you in production or the bad guys exploit them.
+
+**Fuzzing is not a substitute for explicitly testing functional requirements.** Write the appropriate unit/integration/system tests. Then consider additional fuzzing if applicable.
+
+**Fuzz testing must be done continuously.** This problem is solved at Google. Once a fuzz target is submitted, the fuzzing infrastructure will run it 24/7. With corpus growing over time, it provides both intensive testing for the stable code and regression testing for the new changes.
```

Comment:
Probably add a reference to ClusterFuzz somewhere here?

Also mention OSS-Fuzz for in-scope projects (i.e. high impact open source).

Reply:
Done.